File: src/PathTracer.Platform/Platforms/Windows/NativeInputService.cpp
#include "WindowsCommon.h"
#include "../Platform.h"

#define GET_X_LPARAM(lp)    ((int)(short)LOWORD(lp))
#define GET_Y_LPARAM(lp)    ((int)(short)HIWORD(lp))

auto applicationInputQueues = std::map<void*, std::vector<WindowsEvent>>();

void NativeInputProcessKeyboardEvent(void* application, WindowsEvent event) 
{
    if (applicationInputQueues.count(application) == 0) 
    {
        applicationInputQueues[application] = std::vector<WindowsEvent>();
    }

    applicationInputQueues[application].push_back(event);
}

void ProcessKey(NativeInputObject *inputObjects, NativeInputObjectKey key, int windowsKeyCode, WindowsEvent event)
{
    if (event.WParam == windowsKeyCode)
    {
        inputObjects[key].Value = (event.Message == WM_KEYDOWN) ? 1.0f : 0.0f;

        if (event.Message == WM_KEYDOWN)
        {
            if ((event.LParam & 0xFF) == 1)
            {
                inputObjects[key].Repeatcount += 1;
            }
        }
        else
        {
            inputObjects[key].Repeatcount = 0;
        }
    }
}

DllExport void PT_UpdateInputState(void* application, NativeInputState* inputState)
{
    // TODO: Bind correct application
    NativeInputObject* inputObjects = (NativeInputObject*)inputState->InputObjectPointer;

    if (applicationInputQueues.count(nullptr) == 0) 
    {
        return;
    }

    auto events = applicationInputQueues[nullptr];

    for (auto i = 0; i < events.size(); i++)
    {
        auto event = events[i];

        if (event.Message == WM_KEYDOWN || event.Message == WM_KEYUP)
        {
            ProcessKey(inputObjects, NativeInputObjectKey::KeyA, 'A', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyB, 'B', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyC, 'C', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyD, 'D', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyE, 'E', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyF, 'F', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyG, 'G', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyH, 'H', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyI, 'I', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyJ, 'J', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyK, 'K', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyL, 'L', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyM, 'M', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyN, 'N', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyO, 'O', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyP, 'P', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyQ, 'Q', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyR, 'R', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyS, 'S', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyT, 'T', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyU, 'U', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyV, 'V', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyW, 'W', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyX, 'X', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyY, 'Y', event);
            ProcessKey(inputObjects, NativeInputObjectKey::KeyZ, 'Z', event);

        }
        else if (event.Message == WM_LBUTTONDOWN || event.Message == WM_LBUTTONUP)
        {
            inputObjects[NativeInputObjectKey::MouseLeftButton].Value = (event.Message == WM_LBUTTONDOWN) ? 1.0f : 0.0f;
        }
        else if (event.Message == WM_MOUSEMOVE)
        {
            inputObjects[NativeInputObjectKey::MouseAxisX].Value = GET_X_LPARAM(event.LParam);
            inputObjects[NativeInputObjectKey::MouseAxisY].Value = GET_Y_LPARAM(event.LParam);
        }
    }
    
    applicationInputQueues[nullptr].clear();
}

File: src/PathTracer.Platform/Platforms/MacOS/NativeInputService.swift
import Cocoa
import NativeUIServiceModule

var applicationInputQueues = [UnsafeMutablePointer<Int8>: [NSEvent]]()

public func nativeInputProcessEvent(_ application: UnsafeMutablePointer<Int8>, _ event: NSEvent) {
    if (applicationInputQueues[application] == nil) {
        applicationInputQueues[application] = []
    }

    applicationInputQueues[application]!.append(event)
}

@_cdecl("PT_UpdateInputState")
public func updateInputState(application: UnsafeMutablePointer<Int8>, inputState: UnsafeMutablePointer<InputState>) {
    guard let queue = applicationInputQueues[application] else {
        return
    } 

    for event in queue { 
        let inputObjects = inputState.pointee.InputObjectPointer.bindMemory(to: InputObject.self, capacity: Int(inputState.pointee.InputObjectCount))

        if (event.type == .keyUp || event.type == .keyDown) {
            processKey(inputObjects, KeyA, "a", event)
            processKey(inputObjects, KeyB, "b", event)
            processKey(inputObjects, KeyC, "c", event)
            processKey(inputObjects, KeyD, "d", event)
            processKey(inputObjects, KeyE, "e", event)
            processKey(inputObjects, KeyF, "f", event)
            processKey(inputObjects, KeyG, "g", event)
            processKey(inputObjects, KeyH, "h", event)
            processKey(inputObjects, KeyI, "i", event)
            processKey(inputObjects, KeyJ, "j", event)
            processKey(inputObjects, KeyK, "k", event)
            processKey(inputObjects, KeyL, "l", event)
            processKey(inputObjects, KeyM, "m", event)
            processKey(inputObjects, KeyN, "n", event)
            processKey(inputObjects, KeyO, "o", event)
            processKey(inputObjects, KeyP, "p", event)
            processKey(inputObjects, KeyQ, "q", event)
            processKey(inputObjects, KeyR, "r", event)
            processKey(inputObjects, KeyS, "s", event)
            processKey(inputObjects, KeyT, "t", event)
            processKey(inputObjects, KeyU, "u", event)
            processKey(inputObjects, KeyV, "v", event)
            processKey(inputObjects, KeyW, "w", event)
            processKey(inputObjects, KeyX, "x", event)
            processKey(inputObjects, KeyY, "y", event)
            processKey(inputObjects, KeyZ, "z", event)

            // Left Arrow
            processKey(inputObjects, Left, 123, event)
            
            // Right Arrow
            processKey(inputObjects, Right, 124, event)
            
            // Up Arrow
            processKey(inputObjects, Up, 126, event)
            
            // Down Arrow
            processKey(inputObjects, Down, 125, event)

        } else if (event.type == .leftMouseDown || event.type == .leftMouseUp) {
            inputObjects[Int(MouseLeftButton.rawValue)].Value = (event.type == .leftMouseDown) ? 1.0 : 0.0
        } else if (event.type == .mouseMoved || event.type == .leftMouseDragged) {
            let contentView = globalWindow!.window.contentView! as NSView
            let mainScreenScaling = globalWindow!.window.screen!.backingScaleFactor

            let size = contentView.frame.size
            
            inputObjects[Int(MouseAxisX.rawValue)].Value = Float(event.locationInWindow.x) * Float(mainScreenScaling)
            inputObjects[Int(MouseAxisY.rawValue)].Value = (Float(size.height) - Float(event.locationInWindow.y)) * Float(mainScreenScaling)
        }
    }

    applicationInputQueues[application] = []
}

private func processKey(_ inputObjects: UnsafeMutablePointer<InputObject>, _ key: InputObjectKey, _ keyCode: String, _ event: NSEvent) {
   processKey(inputObjects, key, Int(Character(keyCode).asciiValue!), event) 
}

private func processKey(_ inputObjects: UnsafeMutablePointer<InputObject>, _ key: InputObjectKey, _ keyCode: Int, _ event: NSEvent)
{
    var isValid = event.keyCode == keyCode

    if (event.characters != nil) {
        guard let keyChar = event.characters else {
            return
        }

        let asciiValue = Character(keyChar).asciiValue

        if (asciiValue != nil && asciiValue! == keyCode) {
            isValid = true 
        }
    }

    if (isValid) {
        let keyRawValue = Int(key.rawValue)
        inputObjects[keyRawValue].Value = (event.type == .keyDown) ? 1.0 : 0.0;

        if (event.type == .keyDown) {
            inputObjects[keyRawValue].Repeatcount += 1;
        } else {
            inputObjects[keyRawValue].Repeatcount = 0;
        }
    }
}

private func updateInputObject(_ inputObject: inout InputObject, _ event: NSEvent) {
    inputObject.Value = (event.type == .keyDown || event.type == .leftMouseDown) ? 1.0 : 0.0
}
